bluesky/bundlers.py

```diff
@@ -597,68 +631,274 @@ async def complete(self, msg):
+                f"width `{indices_difference}` than other detectors in the same collect() or save()."
+            )
+
+        current_seq_counter = self._sequence_counters[message_stream_name]
+        doc["seq_nums"] = StreamRange(
+            start=current_seq_counter, stop=current_seq_counter + indices_difference
+        )
+
+        return indices_difference
+
+    async def _pack_external_assets(
+        self,
+        asset_docs: Iterable[Tuple[str, ExternalAssetDoc]],
+        message_stream_name: Optional[str]
+    ):
+        """Packs some external asset documents with relevant information from the run."""
+
+        stream_datum_previous_indices_difference = 0
+        number_of_stream_datum_receieved = 0
+
+        for name, doc in asset_docs:
+            if name == DocumentNames.resource.value:
+                doc["run_start"] = self._run_start_uid
+            elif name == DocumentNames.stream_resource.value:
+                if number_of_stream_datum_receieved:
+                    raise RuntimeError("Received a `stream_resource` document after `stream_datum` documents.")
+
+                doc["run_start"] = self._run_start_uid
+
+            elif name == DocumentNames.stream_datum.value:
+                if doc["descriptor"]:
+                    raise RuntimeError(
+                        f"Received a `stream_datum` {doc['uid']} with a `descriptor` uid already "
+                        f"filled in, with the value {doc['descriptor']} this should be an empty string."
+                    )
+
+                if message_stream_name in self._descriptors:
+                    doc["descriptor"] = self._descriptors[message_stream_name].descriptor_doc["uid"]
+                else:
+                    raise RuntimeError(
+                        f"`descriptor` not made for stream {message_stream_name}."
+                    )
+
+                stream_datum_previous_indices_difference = await self._pack_seq_nums_into_stream_datum(
+                    doc,
+                    message_stream_name,
+                    stream_datum_previous_indices_difference
+                )
+                number_of_stream_datum_receieved += 1
+            elif name == DocumentNames.datum.value:
+                ...
+            else:
+                raise RuntimeError(
+                    f"Tried to emit an external asset {name}, acceptable external assets are "
+                    "`resource`, `stream_resource`, `datum`, or `stream_datum`"
+                )
+
+            await self.emit(DocumentNames(name), doc)
+
+            doc_logger.debug(
+                    "[%s] document emitted %r",
+                    name,
+                    doc,
+                    extra={
+                        "doc_name": name,
+                        "run_uid": self._run_start_uid,
+                        "doc": doc
+                    }
+                )
+
+        return stream_datum_previous_indices_difference
```

Comment:
Before we return we should check that we have received a StreamDatum for every `STREAM:` data_key in the descriptor